Improve Primary Key handling for upsert.

lsylvester opened the pull request:
### Summary

Currently when passing in the primary key to the `unique_by` option to upsert, like:

```ruby
Book.upsert_all [{ id: 1, name: new_name }], unique_by: :id
```

You get the error 

```
ArgumentError: No unique index found for id
```

If you don't pass in anything to `unique_by` it does a unique_by on the primary key - so you can work around it - but I don't think that it is clear and that passing in the primary key as a column list should be valid.

Also - the upsert_all code assumes that the `primary_key` for the model is the primary key in the database - which might not always be the case.  

If you have set `primary_key` for a model to something without a unique index and try to upsert, you get an error like:

```
ActiveRecord::StatementInvalid: PG::InvalidColumnReference: ERROR:  there is no unique or exclusion constraint matching the ON CONFLICT specification
```

This change changes this error message to be the same as if it was any set of columns without a unique index - an `ArgumentError` with message like "No unique index found for ..."

lsylvester committed:
handle passing in primary key to unique_by, and handle primary keys missing indexes

kaspth commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -14,7 +14,7 @@ def initialize(model, inserts, on_duplicate:, returning: nil, unique_by: nil)
       @returning = false if @returning == []
 
-      @unique_by = find_unique_index_for(unique_by) if unique_by
+      @unique_by = find_unique_index_for(unique_by || model.primary_key)
```
It's my understanding that it's not always we want to apply a `unique_by` clause.

lsylvester replied:
If the `unique_by` is matching the primary_key - `find_unique_index_for` returns nil - so the update only occurs if the `on_duplicate == :update`.

The reason that is it getting passed in to get the same behaviour for passing in a primary key as you get if you to an `upsert` out specifying the `unique_by`.

This might be better expressed as:

```
unique_by ||= model.primary_key if @on_duplicate == :update
```

And having find_unique_index_for returning a representation of the primary key if the primary key matches.

kaspth commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -32,10 +32,9 @@ def updatable_columns
 
     def primary_keys
-      Array(model.primary_key)
+      Array(connection.schema_cache.primary_keys(model.table_name))
```
```suggestion
      connection.schema_cache.primary_keys(model.table_name)
```

`primary_keys` should return an Array already, iirc.

lsylvester replied:
Looks like having `schema_cache` return an array for `primary_keys` causes issues because array return values get suppressed.

https://github.com/rails/rails/blob/1aa90ddc1c2ae5d44f64f41c6eda3345054d69c2/activerecord/lib/active_record/attribute_methods/primary_key.rb#L92-L93

https://github.com/rails/rails/blob/1aa90ddc1c2ae5d44f64f41c6eda3345054d69c2/activerecord/lib/active_record/attribute_methods/primary_key.rb#L122-L129

kaspth commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -61,6 +60,8 @@ def find_unique_index_for(unique_by)
 
         if index = unique_indexes.find { |i| match.include?(i.name) || i.columns == match }
           index
+        elsif match == primary_keys
```
I don't understand the purpose of this? Also: what happens with different sort orders?

lsylvester replied:
The current implementation also errors if the columns passed in aren't in the same order as the index. I think this requires some further work.

lsylvester replied:
If the primary key is passed in - we don't have a Index object (responding to columns and where) that we could use to generate the sql clause in `conflict_target`. 

Having `unique_by` be nil and `@on_duplicate == :update` was how it said that we should use the primary key. 

Without this condition we get an error if we pass in the primary key to the `unique_by`

kaspth commented on activerecord/test/cases/insert_all_test.rb:
```diff
@@ -225,6 +226,23 @@ def test_upsert_all_updates_existing_records
+
+    new_name = "Agile Web Development with Rails, 4th Edition"
+    Book.upsert_all [{ id: 1, name: new_name }], unique_by: :id
+    assert_equal new_name, Book.find(1).name
```
```suggestion
    Book.upsert_all [{ id: 1, name: "New edition" }], unique_by: :id

    assert_equal "New edition", Book.find(1).name
```

I'd rather we just shorten the text and repeat it than capture it in a local variable.

kaspth commented on activerecord/test/cases/insert_all_test.rb:
```diff
@@ -225,6 +226,23 @@ def test_upsert_all_updates_existing_records
+    skip unless supports_insert_on_duplicate_update?
+
+    error = assert_raises ArgumentError do
+      Speedometer.upsert_all [{ speedometer_id: "s1", name: "New Speesdometer" }]
```
```suggestion
      Speedometer.upsert_all [{ speedometer_id: "s1", name: "New Speedometer" }]
```

lsylvester commented:
I have updated this PR to allow the order of the columns in `unique_by` to not be required to match the order in the index, and refactored to assign a `IndexDefinition` to `@unique_by` when we are matching on the primary key - so that we don't have to handle any special cases when generating the SQL.

kaspth commented:
Merged in 1af44e4aee, thanks @lsylvester ❤️

> the columns in `unique_by` to not be required to match the order in the index

We don't want this, I think it's better that the columns conform to the index order since a different order changes the index cardinality aiui.

But I don't see such a change in the code? It seems your last change was on Jun 25 but your comment was on Jul 5, so maybe there's code missing? Either way, the patch looked shippable as it was now. Thanks!
